doc/md/fundamentals/2-types/3-functions.md

```diff
@@ -0,0 +1,213 @@
+---
+sidebar_position: 3
+---
+
+# Function types
+
+Motoko provides different types of functions based on where in the program they execute and how they interact with the system. Understanding these distinctions is essential when designing efficient and scalable canister logic.
+
+## Function keywords
+
+| Keyword  | Function  |
+|-------------|--------------|
+| `public`    | Makes a function callable by users or other [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters). |
+| `shared`    | Enables [inter-canister](https://internetcomputer.org/docs/motoko/fundamentals/messaging) calls and provides access to the caller’s identity. |
+| `async`     | Runs the function [asynchronously](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) and returns a future result. |
+| `query`     | Optimized for reading data but cannot modify [state](https://internetcomputer.org/docs/motoko/fundamentals/state). |
+
+## Function comparison
+
+| Function type                | Mutates [state](https://internetcomputer.org/docs/motoko/fundamentals/state) | Calls updates | Calls queries | Asynchronous | External calls |
+|------------------------------|---------------|------------------|------------------|---------------|---------------|
+| Local                    | No            | Yes              | Yes              | No            | No            |
+| Generic                  | No            | Yes              | Yes              | No            | No            |
+| Shared update            | Yes           | Yes              | Yes              | Yes           | Yes           |
+| Shared query             | No            | No               | Yes              | Yes           | Yes           |
+| Shared composite query   | No            | No               | Yes              | Yes           | Yes           |
+
+## Local functions
+
+Local functions run within the canister's [actor](https://internetcomputer.org/docs/motoko/fundamentals/actors-async). They do not make calls to other [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters). Local functions are processed fast and synchronously.
+
+```motoko no-repl
+func add(a : Nat, b : Nat) : Nat {
+    a + b
+};
+
+let sum : Nat = add(5, 3); // Synchronous execution
+```
+
+**Example use case:** Simple computations that do not require network calls.
+
+## Generic functions
+
+Generic functions allow the use of type parameters, making them more flexible for using different data types.
+
+```motoko no-repl
+func identity<T>(x : T) : T = x;
+
+let num : Nat = identity<Nat>(42);
+let text : Text = identity<Text>("Hello");
+```
```

Comment:
````suggestion
```

The type of `identity` is `<T>T -> T`, indicating that when supplied with any type `T`, and argument of type `T`, it  returns a result of the same type `T`.

Type arguments can be omitted from calls when the compiler can infer them from the arguments and context, allowing the simpler code:

```motoko no-repl
func identity<T>(x : T) : T = x;
let num : Nat = identity(42);
let text : Text = identity("Hello");
```
````